File: python/unicef_api/flows.py
import pandas as pd
import requests
import xml.etree.ElementTree as ET
import time

def list_dataflows(max_retries: int = 3) -> pd.DataFrame:
    """
    List all available UNICEF SDMX dataflows.
    
    Returns:
        DataFrame with columns: id, name, agency, version
    
    Example:
        >>> from unicef_api import list_dataflows
        >>> flows = list_dataflows()
        >>> print(flows.head())
    """
    
    url = "https://sdmx.data.unicef.org/ws/public/sdmxapi/rest/dataflow/UNICEF?references=none&detail=full"
    
    for attempt in range(max_retries):
        try:
            response = requests.get(url, timeout=60)
            response.raise_for_status()
            
            # Parse XML response
            root = ET.fromstring(response.content)
            
            # Extract dataflows
            ns = {'s': 'http://www.sdmx.org/resources/sdmxml/schemas/v2_1/structure'}
            dataflows = []
            
            for df in root.findall('.//s:Dataflow', ns):
                name_elem = df.find('.//s:Name', ns)
                dataflows.append({
                    'id': df.get('id'),
                    'agency': df.get('agencyID'),
                    'version': df.get('version'),
                    'name': name_elem.text if name_elem is not None else ''
                })
            
            return pd.DataFrame(dataflows)
            
        except Exception as e:
            if attempt == max_retries - 1:
                raise
            time.sleep(1)
    
    return pd.DataFrame()


def dataflow_schema(dataflow: str, metadata_dir: str = None) -> dict:
    """
    Get dataflow schema information (dimensions and attributes).
    
    Reads from local YAML schema files in metadata/current/dataflows/.
    
    Args:
        dataflow: The dataflow ID (e.g., "CME", "EDUCATION").
        metadata_dir: Optional path to metadata directory. Auto-detected if None.
    
    Returns:
        Dictionary with keys: id, name, version, agency, dimensions, attributes,
        time_dimension, primary_measure.
    
    Example:
        >>> from unicef_api import dataflow_schema
        >>> schema = dataflow_schema("CME")
        >>> print(schema['dimensions'])
        ['REF_AREA', 'INDICATOR', 'SEX', 'WEALTH_QUINTILE']
        >>> print(schema['attributes'])
        ['DATA_SOURCE', 'COUNTRY_NOTES', 'REF_PERIOD', ...]
    """
    import yaml
    from pathlib import Path
    import os
    
    df_upper = dataflow.upper()
    
    # Find metadata directory
    if metadata_dir is None:
        metadata_dir = _find_metadata_dir()
    
    metadata_path = Path(metadata_dir)
    schema_path = metadata_path / "dataflows" / f"{df_upper}.yaml"
    
    if not schema_path.exists():
        # Fall back to basic info from _unicefdata_dataflows.yaml
        basic = _get_basic_dataflow_info(df_upper, metadata_path)
        if basic:
            print(f"Note: Detailed schema not available for '{df_upper}'. Showing basic info.")
            return basic
        raise FileNotFoundError(
            f"Dataflow '{df_upper}' not found. Use list_dataflows() to see available dataflows."
        )
    
    # Parse YAML schema
    with open(schema_path, 'r', encoding='utf-8') as f:
        schema = yaml.safe_load(f)
    
    # Extract dimensions (list of id values)
    dimensions = []
    if schema.get('dimensions'):
        dimensions = [d.get('id', '') for d in schema['dimensions']]
    
    # Extract attributes (list of id values)
    attributes = []
    if schema.get('attributes'):
        attributes = [a.get('id', '') for a in schema['attributes']]
    
    return {
        'id': schema.get('id', df_upper),
        'name': schema.get('name', ''),
        'version': schema.get('version', ''),
        'agency': schema.get('agency', 'UNICEF'),
        'dimensions': dimensions,
        'attributes': attributes,
        'time_dimension': schema.get('time_dimension', 'TIME_PERIOD'),
        'primary_measure': schema.get('primary_measure', 'OBS_VALUE'),
    }


def print_dataflow_schema(schema: dict) -> None:
    """
    Pretty-print a dataflow schema.
    
    Args:
        schema: Dictionary from dataflow_schema().
    
    Example:
        >>> schema = dataflow_schema("CME")
        >>> print_dataflow_schema(schema)
    """
    print()
    print("-" * 70)
    print(f"Dataflow Schema: {schema['id']}")
    print("-" * 70)
    print()
    
    if schema.get('name'):
        print(f"Name: {schema['name']}")
    if schema.get('version'):
        print(f"Version: {schema['version']}")
    if schema.get('agency'):
        print(f"Agency: {schema['agency']}")
    print()
    
    dims = schema.get('dimensions', [])
    if dims:
        print(f"Dimensions ({len(dims)}):")
        for d in dims:
            print(f"  {d}")
        print()
    
    attrs = schema.get('attributes', [])
    if attrs:
        print(f"Attributes ({len(attrs)}):")
        for a in attrs:
            print(f"  {a}")
    
    print()
    print("-" * 70)


def _find_metadata_dir() -> str:
    """Find metadata directory. Returns path as string."""
    from pathlib import Path
    import os
    
    # 1. Environment override
    env_home = os.environ.get('UNICEF_DATA_HOME_PYTHON') or os.environ.get('UNICEF_DATA_HOME', '')
    if env_home:
        metadata_dir = Path(env_home) / "metadata" / "current"
        if metadata_dir.exists():
            return str(metadata_dir)
    
    # 2. Relative to module location
    module_dir = Path(__file__).parent
    candidates = [
        module_dir / "metadata" / "current",           # unicef_api/metadata/current
        module_dir.parent / "metadata" / "current",   # python/metadata/current
    ]
    for path in candidates:
        if path.exists():
            return str(path.resolve())
    
    # 3. User home directory
    home_dir = Path.home() / ".unicef_data" / "python" / "metadata" / "current"
    if home_dir.exists():
        return str(home_dir)
    
    raise FileNotFoundError("Could not find metadata directory. Run sync_metadata() first.")


def _get_basic_dataflow_info(dataflow: str, metadata_path) -> dict:
    """Get basic dataflow info from _unicefdata_dataflows.yaml."""
    import yaml
    from pathlib import Path
    
    df_file = Path(metadata_path) / "_unicefdata_dataflows.yaml"
    if not df_file.exists():
        return None
    
    with open(df_file, 'r', encoding='utf-8') as f:
        all_flows = yaml.safe_load(f)
    
    if dataflow in all_flows:
        info = all_flows[dataflow]
        return {
            'id': dataflow,
            'name': info.get('name', ''),
            'version': info.get('version', ''),
            'agency': 'UNICEF',
            'dimensions': [],
            'attributes': [],
        }
    return None

File: R/flows.R
# R/flows.R

#' @title List available SDMX “flows” for an agency
#' @description Download and cache the SDMX dataflow definitions from a specified agency's REST endpoint.
#' @param agency Character agency ID (e.g., "UNICEF").
#' @param retry Number of retries for transient HTTP failures; default is 3.
#' @param cache_dir Directory for on-disk cache; created if it does not exist.
#' @return A tibble with columns `id`, `agency`, `version`, and `name`.
#' @export
#' @importFrom xml2 read_xml xml_find_all xml_attr xml_find_first xml_text
#' @importFrom tibble tibble
#' @importFrom memoise memoise cache_filesystem
#' @importFrom tools R_user_dir
list_sdmx_flows <- local({
  memo_env <- new.env(parent = emptyenv())

  fetch_flows <- function(
    agency    = "UNICEF",
    retry     = 3L,
    cache_dir = tools::R_user_dir("unicefdata", "cache")
  ) {
    if (!dir.exists(cache_dir)) dir.create(cache_dir, recursive = TRUE)

    url <- sprintf(
      "https://sdmx.data.unicef.org/ws/public/sdmxapi/rest/dataflow/%s?references=none&detail=full",
      agency
    )

    xml_text <- .fetch_sdmx(url, retry = retry)
    doc <- xml2::read_xml(xml_text)
    dfs <- xml2::xml_find_all(doc, ".//str:Dataflow")

    tibble::tibble(
      id      = xml2::xml_attr(dfs, "id"),
      agency  = xml2::xml_attr(dfs, "agencyID"),
      version = xml2::xml_attr(dfs, "version"),
      name    = xml2::xml_text(
        xml2::xml_find_first(dfs, "./com:Name[@xml:lang='en']")
      )
    )
  }

  get_memoised <- function(cache_dir) {
    key <- normalizePath(cache_dir, winslash = "/", mustWork = FALSE)
    if (!nzchar(key)) key <- cache_dir

    if (!exists(key, envir = memo_env, inherits = FALSE)) {
      cache <- memoise::cache_filesystem(cache_dir)
      assign(key, memoise::memoise(fetch_flows, cache = cache), envir = memo_env)
    }

    get(key, envir = memo_env, inherits = FALSE)
  }

  function(
    agency    = "UNICEF",
    retry     = 3L,
    cache_dir = tools::R_user_dir("unicefdata", "cache")
  ) {
    memoised_fn <- get_memoised(cache_dir)
    memoised_fn(agency = agency, retry = retry, cache_dir = cache_dir)
  }
})


#' @title Get dataflow schema information
#' @description Display the dimensions and attributes for a UNICEF dataflow.
#'   Reads from local YAML schema files in metadata/current/dataflows/.
#' @param dataflow Character. The dataflow ID (e.g., "CME", "EDUCATION").
#' @param metadata_dir Optional path to metadata directory. Auto-detected if NULL.
#' @return A list with components: id, name, version, agency, dimensions, attributes.
#' @export
#' @examples
#' \dontrun{
#' # Get schema for Child Mortality dataflow
#' schema <- dataflow_schema("CME")
#' print(schema$dimensions)
#' print(schema$attributes)
#' }
dataflow_schema <- function(dataflow, metadata_dir = NULL) {

  # Null coalescing operator
  `%||%` <- function(x, y) if (is.null(x)) y else x

  # Convert to uppercase
  df_upper <- toupper(dataflow)

  # Find metadata directory
  if (is.null(metadata_dir)) {
    metadata_dir <- .find_metadata_dir()
  }

  # Look for schema file
  schema_path <- file.path(metadata_dir, "dataflows", paste0(df_upper, ".yaml"))

  if (!file.exists(schema_path)) {
    # Fall back to _unicefdata_dataflows.yaml for basic info
    basic <- .get_basic_dataflow_info(df_upper, metadata_dir)
    if (!is.null(basic)) {
      message(sprintf("Note: Detailed schema not available for '%s'. Showing basic info.", df_upper))
      return(basic)
    }
    stop(sprintf("Dataflow '%s' not found. Use list_sdmx_flows() to see available dataflows.", df_upper))
  }

  # Parse YAML schema
  schema <- yaml::read_yaml(schema_path)

  # Extract dimensions (list of id values)
  dimensions <- if (!is.null(schema$dimensions)) {
    sapply(schema$dimensions, function(d) d$id)
  } else {
    character(0)
  }

  # Extract attributes (list of id values)
  attributes <- if (!is.null(schema$attributes)) {
    sapply(schema$attributes, function(a) a$id)
  } else {
    character(0)
  }

  result <- list(
    id = schema$id %||% df_upper,
    name = schema$name %||% "",
    version = schema$version %||% "",
    agency = schema$agency %||% "UNICEF",
    dimensions = dimensions,
    attributes = attributes,
    time_dimension = schema$time_dimension %||% "TIME_PERIOD",
    primary_measure = schema$primary_measure %||% "OBS_VALUE"
  )

  class(result) <- c("unicef_dataflow_schema", "list")
  result
}


#' Print method for dataflow schema
#' @param x A unicef_dataflow_schema object
#' @param ... Additional arguments (ignored)
#' @export
print.unicef_dataflow_schema <- function(x, ...) {
  cat("\n")
  cat(strrep("-", 70), "\n")
  cat("Dataflow Schema:", x$id, "\n")
  cat(strrep("-", 70), "\n")
  cat("\n")

  if (nzchar(x$name)) cat("Name:", x$name, "\n")
  if (nzchar(x$version)) cat("Version:", x$version, "\n")
  if (nzchar(x$agency)) cat("Agency:", x$agency, "\n")
  cat("\n")

  if (length(x$dimensions) > 0) {
    cat("Dimensions (", length(x$dimensions), "):\n", sep = "")
    for (d in x$dimensions) {
      cat("  ", d, "\n", sep = "")
    }
    cat("\n")
  }

  if (length(x$attributes) > 0) {
    cat("Attributes (", length(x$attributes), "):\n", sep = "")
    for (a in x$attributes) {
      cat("  ", a, "\n", sep = "")
    }
  }

  cat("\n")
  cat(strrep("-", 70), "\n")
  invisible(x)
}


#' Find metadata directory
#' @keywords internal
.find_metadata_dir <- function() {
  # 1. Environment override
  env_home <- Sys.getenv("UNICEF_DATA_HOME_R", Sys.getenv("UNICEF_DATA_HOME", ""))
  if (nzchar(env_home)) {
    metadata_dir <- file.path(env_home, "metadata", "current")
    if (dir.exists(metadata_dir)) return(metadata_dir)
  }

  # 2. Relative to working directory
  script_dir <- getwd()
  candidates <- c(
    file.path(script_dir, "R", "metadata", "current"),
    file.path(script_dir, "metadata", "current"),
    file.path(script_dir, "..", "R", "metadata", "current"),
    file.path(script_dir, "..", "metadata", "current")
  )
  for (path in candidates) {
    if (dir.exists(path)) return(normalizePath(path))
  }

  # 3. User cache directory
  if (exists("R_user_dir", envir = asNamespace("tools"))) {
    base_dir <- tryCatch(tools::R_user_dir("unicefdata", "cache"), error = function(e) "")
    if (nzchar(base_dir)) {
      metadata_dir <- file.path(base_dir, "metadata", "current")
      if (dir.exists(metadata_dir)) return(metadata_dir)
    }
  }

  # 4. Home directory fallback
  home_dir <- file.path(Sys.getenv("HOME"), ".unicef_data", "r", "metadata", "current")
  if (dir.exists(home_dir)) return(home_dir)

  stop("Could not find metadata directory. Run sync_metadata() first.")
}


#' Get basic dataflow info from _unicefdata_dataflows.yaml
#' @keywords internal
.get_basic_dataflow_info <- function(dataflow, metadata_dir) {
  `%||%` <- function(x, y) if (is.null(x)) y else x

  df_file <- file.path(metadata_dir, "_unicefdata_dataflows.yaml")
  if (!file.exists(df_file)) return(NULL)

  all_flows <- yaml::read_yaml(df_file)
  if (dataflow %in% names(all_flows)) {
    info <- all_flows[[dataflow]]
    return(list(
      id = dataflow,
      name = info$name %||% "",
      version = info$version %||% "",
      agency = "UNICEF",
      dimensions = character(0),
      attributes = character(0)
    ))
  }
  NULL
}
